packages/docusaurus-theme-common/src/contexts/tabGroupChoice.tsx

```diff
@@ -13,61 +13,123 @@ import React, {
+    newChoice: string,
+  ) => void;
 };
 
 const Context = React.createContext<ContextValue | undefined>(undefined);
 
+function getSearchKey(
+  groupId: string | undefined,
+  queryString: string | boolean | undefined,
+) {
+  if (typeof queryString === 'string') {
+    return queryString;
+  }
+  if (queryString === false) {
+    return undefined;
+  }
+  if (queryString === true && !groupId) {
+    throw new Error(
+      `Docusaurus error: The <Tabs> component groupId prop is required if queryString=true, because this value is used as the search param name. You can also provide an explicit value such as queryString="my-search-param".`,
+    );
+  }
+  return groupId;
+}
+
 function useContextValue(): ContextValue {
-  const [ready, setReady] = useState(false);
-  const [tabGroupChoices, setChoices] = useState<{
-    readonly [groupId: string]: string;
-  }>({});
-  const setChoiceSyncWithLocalStorage = useCallback(
+  // TODO not re-render optimized
+  // See https://thisweekinreact.com/articles/useSyncExternalStore-the-underrated-react-api
+  const location = useLocation();
+  const history = useHistory();
+
+  const [tabGroupChoicesInStorage, setGroupChoicesInStorage] = useState<
+    ContextValue['tabGroupChoicesInStorage']
+  >({});
+  const [tabGroupChoicesInQueryParams, setGroupChoicesInQueryParams] = useState<
+    ContextValue['tabGroupChoicesInQueryParams']
+  >({});
+
+  const updateLocalStorage = useCallback(
     (groupId: string, newChoice: string) => {
       createStorageSlot(`${TAB_CHOICE_PREFIX}${groupId}`).set(newChoice);
     },
     [],
   );
+  const updateHistory = useCallback(
+    (searchKey: string, newTabValue: string) => {
+      const searchParams = new URLSearchParams(location.search);
+      searchParams.set(searchKey, newTabValue);
+      history.replace({...location, search: searchParams.toString()});
+    },
+    [history, location],
+  );
 
   useEffect(() => {
     try {
       const localStorageChoices: {[groupId: string]: string} = {};
       listStorageKeys().forEach((storageKey) => {
         if (storageKey.startsWith(TAB_CHOICE_PREFIX)) {
-          const groupId = storageKey.substring(TAB_CHOICE_PREFIX.length);
-          localStorageChoices[groupId] = createStorageSlot(storageKey).get()!;
+          const groupIdFromStorage = storageKey.substring(
+            TAB_CHOICE_PREFIX.length,
+          );
+          localStorageChoices[groupIdFromStorage] =
+            createStorageSlot(storageKey).get()!;
         }
       });
-      setChoices(localStorageChoices);
+      setGroupChoicesInStorage(localStorageChoices);
     } catch (err) {
       console.error(err);
     }
-    setReady(true);
   }, []);
 
-  const setTabGroupChoices = useCallback(
-    (groupId: string, newChoice: string) => {
-      setChoices((oldChoices) => ({...oldChoices, [groupId]: newChoice}));
-      setChoiceSyncWithLocalStorage(groupId, newChoice);
+  const setTabGroupChoice = useCallback(
```

Comment:
`setTabGroupChoice` pushes tab choice to local storage or query param depending on `groupId` & `queryString`.